File: algorithms/leetcode/maximum_average_subarray_1.py


class Solution:
    def findMaxAverage(self, nums):
        """
        You are given an integer array nums consisting of n elements, and an integer k.
        Find a contiguous subarray whose length is equal to k that has the maximum average value and return this value. Any answer with a calculation error less than 10-5 will be accepted.


        Example 1:
        =======================================================================
        Input: nums = [1,12,-5,-6,50,3], k = 4
        Output: 12.75000
        Explanation: Maximum average is (12 - 5 - 6 + 50) / 4 = 51 / 4 = 12.75
        =======================================================================

        """

        left = 1 
        right = k
        max_sum = sum(nums[:k])
        print(max_sum)
        current_sum = max_sum 
        while right  < len(nums):
            current_sum = current_sum - nums[left-1] + nums[right]

            max_sum = max(max_sum,current_sum)
            left +=1
            right +=1
            

        return max_sum / k 
    



if __name__ == "__main__":

    nums = [1,12,-5,-6,50,3]
    k = 4 
    solution = Solution()
    res = solution.findMaxAverage(nums)
    print(res)

File: algorithms/leetcode/maximum_average_subarray_1.cpp
#include <vector>
#include <iostream>
double max(double a, double b) {
    if (a > b) return a;
    return b;
}

double findMaxAverage(std::vector<int>& nums, int k) {
    double max_sum;
    for (int i = 0; i < k; i +=1) {
      max_sum += (double)nums[i];
    }
    double current_sum = max_sum;
    int left = 1;
    int right = k;
    while (right < nums.size()) {
      current_sum = (double)(current_sum - nums[left-1] + nums[right]);
      max_sum = max(max_sum, current_sum);
      right +=1;
      left +=1;
    }
    return (double)(max_sum / k);
}


int main() {
  std::vector<int> nums = {1,12,-5,-6,50,3};
  int k = 4;
  double res = findMaxAverage(nums, k);
  std::cout << res << "\n";
}
